design.md

```diff
@@ -0,0 +1,306 @@
+3. When the **parent** creates the **child** `os/exec Cmd` a context cancellation function will be stored as part of the `Job`. Any process seeking to stop a `Job` may call the cancellation function. This will kill both the **child** and **grandchild**.
+4. The Cgroup related to `Job` will also be removed in unrelated logic.
+
+##### Status
+
+The status of underlying **grandchild** will be returned through the **child** to **parent** as exitcodes. This status will be handling in step 9 of [Starting](#starting).
+
+###### JobStatus
+
+The JobStatus type indicates the status of a job. The status will be one of the following:
+
+- *running*: Job has been started and is currently running.
+- *stopped*: Job has been forcibly stopped by jobworker.
+- *exited*: Job has exited. (with exit code)
+
+##### Streaming Output
+
+Streaming output will involve reading a log file from `/var/log/jobworker` in chunks and writing it to the client until the client ends the stream.
+
+###### Implementation
+
+When an output stream is requested, the following will occur:
+1. `Job` will be fetched by `job.Service`
+2. `Job.OutputStream` will be called in its own goroutine and accept `context.Context` and `chan<- byte` arguments.
+3. `Job.OutputStream` will open the log file, and defer closing the fd.
+4. `Job.OutputStream` will start a goroutine listening on `<-ctx.Done()`.
+5. `Job.OutputStream` will read from log file, in chunks.
+6. When `Read` returns `io.EOF`, the process will sleep for a second and continue reading. See [Streaming Tradoffs](#streaming-tradeoffs) below.
+7. If the `context.Context` is cancelled, `job.OutputStream` will close the `io.ReaderCloser`, close the `chan<- byte`, and return.
+8. Meanwhile, `grpc.Service.Output()` is streaming these log chunks to a client.
+
+###### Streaming Tradeoffs
+
+Step 6 specifies how to handle `io.EOF`, and I don't really like the approach of polling the file to see if more data is available. However, I don't want to invest effort in setting up a filewatcher for this challenge if it's not necessary.
+
+#### Types
+
+```
+type Service struct {
+  ID            uuid.UUID
+}
+
+func (s Service) StartJob(job Job) error
+func (s Service) StopJob(job Job) error
+func (s Service) FetchJob(jobID uuid.UUID) (*Job, error)
+
+type Job struct {
+  ID      uuid.UUID
+  command *exec.Cmd
+  status  JobStatus
+  output  io.ReadWriterCloser
+
+  user string
+  cgroupID uuid.UUID
+
+// Locking mechanisms not included for brevity.
+// ...
+}
+
+func (j Job) Output() []byte
+func (j Job) StreamOutput(ctx context.Context) (<-chan byte, error)
+```
+---
+
+### grpc
+
+Package grpc will provide a `JobWorker` type that implements the gRPC server stub.
+
+The `JobWorker` type will be responsible for the following.
+- Ensuring all requests have the necessary permissions to be carried out.
+- Utilize `job.Service` and `cgroups.Service` in coordination to satisfy requests.
```

Comment:
Should we push cgroups handling entirely to job.Service? Or do you have particular actions in mind for the gRPC to do?

Reply:
Now that I've adjusted the #Starting and #Stopping sections, I do think it makes sense for `cgroups.Service` to be called by `job.Service`. It will allow for the utilization of cgroups to be abstracted form the `grpc.JobWorker` type and should increase readability of my implementation. I'll update the docs.

Reply:
Updated.